build-script-helper.py

```diff
@@ -60,44 +64,100 @@ def run(args):
-    generate_xcodeproj(args.package_dir,
-      swift_exec=args.swift_exec,
-      sourcekit_searchpath=sourcekit_searchpath,
-      verbose=args.verbose)
-
-  if "test" in args.build_actions or "all" in args.build_actions:
+    try:
+      generate_xcodeproj(args.package_dir,
+        swift_exec=args.swift_exec,
+        sourcekit_searchpath=sourcekit_searchpath,
+        verbose=args.verbose)
+    except subprocess.CalledProcessError as e:
+      printerr('FAIL: Generating the Xcode project failed')
+      printerr('Executing: %s' % ' '.join(e.cmd))
+      sys.exit(1)
+
+  if should_run_action("test", args.build_actions):
     print("** Testing %s **" % package_name)
-    invoke_swift(package_dir=args.package_dir,
-      swift_exec=args.swift_exec,
-      action='test',
-      sourcekit_searchpath=sourcekit_searchpath,
-      # note: test uses a different build_dir so it doesn't interfere with the 'build' step's products before install
-      build_dir=os.path.join(args.build_dir, 'test-build'),
-      config='debug',
-      verbose=args.verbose)
-
-  if "install" in args.build_actions or "all" in args.build_actions:
+    try:
+      invoke_swift(package_dir=args.package_dir,
+        swift_exec=args.swift_exec,
+        action='test',
+        sourcekit_searchpath=sourcekit_searchpath,
+        # note: test uses a different build_dir so it doesn't interfere with the 'build' step's products before install
+        build_dir=os.path.join(args.build_dir, 'test-build'),
+        config='debug',
+        verbose=args.verbose)
+    except subprocess.CalledProcessError as e:
+      printerr('FAIL: Testing %s failed' % package_name)
+      printerr('Executing: %s' % ' '.join(e.cmd))
+      sys.exit(1)
+
+  if should_run_action("install", args.build_actions):
     print("** Installing %s **" % package_name)
     stdlib_dir = os.path.join(args.toolchain, 'usr', 'lib', 'swift', 'macosx')
-    install_package(args.package_dir,
-      install_dir=args.prefix,
-      sourcekit_searchpath=sourcekit_searchpath,
-      build_dir=output_dir,
-      rpaths_to_delete=[stdlib_dir],
-      verbose=args.verbose)
+    try:
+      install_package(args.package_dir,
+        install_dir=args.prefix,
+        sourcekit_searchpath=sourcekit_searchpath,
+        build_dir=output_dir,
+        rpaths_to_delete=[stdlib_dir],
+        verbose=args.verbose)
+    except subprocess.CalledProcessError as e:
+      printerr('FAIL: Installing %s failed' % package_name)
+      printerr('Executing: %s' % ' '.join(e.cmd))
+      sys.exit(1)
+
+
+# Returns true if any of the actions in `action_names` should be run.
+def should_run_any_action(action_names, selected_actions):
+  for action_name in action_names:
+    if should_run_action(action_name, selected_actions):
+      return True
+  return False
+
+
+def should_run_action(action_name, selected_actions):
+  if action_name in selected_actions:
+    return True
+  elif "all" in selected_actions:
```

Comment:
Not important, but you could just `return "all" in selected_actions` at this point, or replace the whole body with something like `return any(item in (action_name, "all") for item in selected_actions)`.

Reply:
I actually find it clearer this way. The second suggestion means, I actually need to think about what’s going on and I like repeating the same pattern that I used for `action_name` already for `"all"`. It’s a bit more verbose but I find it reads clearer.

Reply:
No worries!